transforms/string-concat-to-template.js

```diff
@@ -0,0 +1,115 @@
+'use strict';
+
+const helpers = require('./common/helpers');
+
+
+module.exports = function(file, api, options) {
+  const j = api.jscodeshift;
+  const root = j(file.source);
+
+  const concatenations = root.find(j.Literal)
+      .filter(p => typeof p.value.value === 'string')
+      .filter(p => isConcatenation(p.parent.value))
+      .map(getTopConcatBinaryExpression)
+      .filter(p => hasNonStringLiteral(p.value));
+
+  if (concatenations.size() === 0) {
+    return null;
+  }
+
+  concatenations.replaceWith(p => templateLiteralFromBinaryExpressions(j, p));
+  return root.toSource({ quote: 'single' });
+};
+
+
+function templateLiteralFromBinaryExpressions(j, path) {
+  const nodes = prepareTemplateNodes(j, path);
+  const quasis = [];
+  const expressions = [];
+  const last_index = nodes.length - 1;
+
+  nodes.forEach((node, i) => {
+    const is_quasi = !(i % 2);
+    if (is_quasi) {
+      const value = { cooked: node.value, raw: node.value };
+      const tail = i === last_index;
+      quasis.push(j.templateElement(value, tail));
+    }
+    else {
+      expressions.push(node);
+    }
+  });
+
+  return j.templateLiteral(quasis, expressions);
+}
+
+
+function prepareTemplateNodes(j, path) {
+  const nodes = [];
+  let last_node_is_str = false;
+
+  inorderTreeTraversal(path.value, node => {
+    const is_str = isStringLiteral(node);
+
+    if (nodes.length === 0 && !is_str) {
+      last_node_is_str = true;
+      nodes.push(j.literal(''));
+    }
+
+    if (is_str && last_node_is_str) {
+      const last = nodes.pop();
+      nodes.push(j.literal(last.value + node.value));
+    }
+    else {
+      last_node_is_str = is_str;
+      nodes.push(node);
+    }
+  });
+
+  if (!last_node_is_str) {
+    nodes.push(j.literal(''));
+  }
+  return nodes;
+}
+
+
+// In-order depth-first traversal of the BinaryExpression tree.
+// The visit function is only called with leaf nodes.
+function inorderTreeTraversal(node, visit) {
+  const leaf = node.type !== 'BinaryExpression';
+  if (leaf) {
+    visit(node);
+  }
+  else {
+    inorderTreeTraversal(node.left, visit);
+    inorderTreeTraversal(node.right, visit);
+  }
+}
+
+
+function isConcatenation(node) {
+  return !!node && node.type === 'BinaryExpression' && node.operator === '+';
```

Comment:
- [ ] `isConcatenation` should check that at least one of `left` or `right` is a string literal
- [ ] Rename to `isStringConcatenation`